### PR TITLE
Refactoring, add more ZK checks

### PR DESCRIPTION
**Pull request by avnik**

### CODE REVIEW

File: checks/zk_quorum.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+package checks
+
+import (
+    "bufio"
+    "fmt"
+	"time"
```
Comment (siddharthist) [inline checks/zk_quorum.go]: Looks like some indentation issues here. All code should be run through `gofmt`.

File: checks/zk_quorum.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+func init() { 
+    chkutil.Register("ZooKeeperQuorum", func() chkutil.Check {
+        return &ZooKeeperQuorum{
+            Config: "/etc/zookeeper/conf/zoo.cfg",  /* default value, only for new k/v arg parsing */
+        }
+    })
+}
+
+func (chk ZooKeeperQuorum) New(params []string) (chkutil.Check, error) {
+	if len(params) != 2 {
```
Comment (siddharthist) [inline checks/zk_quorum.go]: If we check that the number of parameters is exactly 2, we lose the benifit of having a default value (line 34). Maybe if there's just one, we assume it's the timeout? I like the idea of being able to specify default values).

File: checks/zk_quorum.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            servers = append(servers, fmt.Sprintf("%s:%s", host, port))
+        }
+    }
+    return servers, nil
+}
+
+func (chk ZooKeeperQuorum) Status() (int, string, error) {
+    servers, err := chk.LoadConfig()
+    if err != nil {
+        return 1, "Failed: " + err.Error(), nil
```
Comment (siddharthist) [inline checks/zk_quorum.go]: Maybe a more descriptive message, like "Failed to load ZK config"?

File: checks/zookeeper.go [hunk 4]
```diff
@@ -55,3 +65,65 @@ func (chk ZooKeeperRUOK) Status() (int, string, error) {
...
+	// match zookeeper servers with failures for error message
+	for i, ok := range oks {
+		if ok == nil {
+            failed = append(failed, fmt.Sprintf("%s: failed to connect", chk.servers[i]))
+        }
+        if ok.MinLatency > chk.minLatency {
+            failed = append(failed, fmt.Sprintf("%s: min latency too big: %v", chk.servers[i], ok.MinLatency)) 
+        }
+        if ok.MaxLatency > chk.maxLatency {
+            failed = append(failed, fmt.Sprintf("%s: min latency too big: %v", chk.servers[i], ok.MaxLatency))
```
Comment (siddharthist) [inline checks/zookeeper.go]: This message doesn't match the condition

File: checks/zookeeper.go [hunk 5]
```diff
@@ -55,3 +65,65 @@ func (chk ZooKeeperRUOK) Status() (int, string, error) {
...
+	servers []string
+    minLatency int64
+    maxLatency int64
+    avgLatency int64
+}
+
+
+func (chk ZooKeeperServerStats) New(params []string) (chkutil.Check, error) {
+	if len(params) < 5 {
+		return chk, errutil.ParameterLengthError{2, params}
```
Comment (siddharthist) [inline checks/zookeeper.go]: 2 -> 5

File: checks/zookeeper.go [hunk 6]
```diff
@@ -55,3 +65,65 @@ func (chk ZooKeeperRUOK) Status() (int, string, error) {
 	}
 	return 1, "Failed: " + failed, nil
 }
+
+type ZooKeeperServerStats struct {
+	timeout time.Duration
+	servers []string
```
Comment (siddharthist) [inline checks/zookeeper.go]: Again, `gofmt`

File: checks/fixtures/zoo_test.cfg [hunk 7]
```diff
@@ -0,0 +1,23 @@
+maxClientCnxns=200
```
Comment (siddharthist) [inline checks/fixtures/zoo_test.cfg]: Maybe this could be in a subdirectory named "test"? I wouldn't know just looking at it what a "fixture" was.

File: checks/zk_quorum.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            default:
+                unknowns++
+            }
+        } else {
+            unknowns++
+        }
+    }
+
+    if leaders != 1 {
+        failed = append(failed, fmt.Sprintf("Cluster have more than one leader"))
```
Comment (siddharthist) [inline checks/zk_quorum.go]: This is minor, but "have" -> "has"

File: checks/zookeeper.go [hunk 9]
```diff
@@ -55,3 +65,65 @@ func (chk ZooKeeperRUOK) Status() (int, string, error) {
...
+    minLatency int64
+    maxLatency int64
+    avgLatency int64
+}
+
```
Comment (siddharthist) [inline checks/zookeeper.go]: Can we get a documentation comment here? Should specify units for the latencies.

File: checks/zookeeper_test.go [hunk 10]
```diff
@@ -1,7 +1,8 @@
-package checks
+package checks_test
```
Comment (siddharthist) [inline checks/zookeeper_test.go]: These do need to be in the "checks" package. From [the golang documentation](https://golang.org/pkg/testing/),

> To write a new test suite, create a file whose name ends _test.go that contains the TestXxx functions as described here. Put the file in the same package as the one being tested. The file will be excluded from regular package builds but will be included when the “go test” command is run. For more detail, run “go help test” and “go help testflag”.

File: checks/zk_quorum.go (hunk 1 above)
Comment (avnik) [inline checks/zk_quorum.go]: Agreed. Will format this concrete line by hands, then run `go fmt` after merge this request (we have a lot of unformatted code, and it just smash PR with unrelated changes)

File: checks/zk_quorum.go (hunk 2 above)
Comment (avnik) [inline checks/zk_quorum.go]: Defaults will work only with new K/V arg parsing, it just some residue from ruther development

File: checks/zk_quorum.go (hunk 3 above)
Comment (avnik) [inline checks/zk_quorum.go]: agreed

File: checks/fixtures/zoo_test.cfg (hunk 7 above)
Comment (avnik) [inline checks/fixtures/zoo_test.cfg]: fixtures/  just to avoid mess test datafiles (configs) with code

File: checks/zookeeper_test.go (hunk 10 above)
Comment (avnik) [inline checks/zookeeper_test.go]: It a common practice (for example -- https://github.com/golang/go/blob/master/src/strings/reader_test.go), But I postpone this one commit, and we can discuss it later (It not essential ;))

File: checks/zookeeper.go (hunk 5 above)
Comment (avnik) [inline checks/zookeeper.go]: Is bit rewritten your code,  4 positional parameters, + server list (as in RUOK module)

Comment (avnik) [inline checks/zookeeper.go]: Is bit rewritten your code,  4 positional parameters, + server list (as in RUOK module)

File: checks/zk_quorum_test.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package checks
+
+import (
+    "testing"
+//    "github.com/CiscoCloud/distributive/checks"
```
Comment (siddharthist) [inline checks/zk_quorum_test.go]: Can we remove this?

**Comment by siddharthist** [comment]
LGTM :+1:
